Abstract vision framework with included support for focus, navigator object and browse mode caret highlighting

LeonarddeR opened the pull request:
Pr provided on behalf of @babbagecom

### Link to issue number
Implements #971, but without a gui

### Background
For years, NVDA has been an open source screen reader for the blind and partially sighted people who rely on braille and speech output. There are no built-in facilities within NVDA to manipulate the contents of the screen. There are several outstandign issues asking for such features, such as:

* #7857: Making the screen black (i.e. a screen curtain) while NVDA is active, mainly for privacy reasons
* #971: Visual highlight of focus, review or browse mode caret location

There is also a long standing desire to have a basic screen magnification facility within NVDA.

This pr intends to lay the base of a vision framework that can be used to implement such functionality in the core of NVDA. Though there is no GUI in the current pull request, the framework is fully working. The GUI will be added in a follow up pr.

### A new vision module
Most magic is supposed to happen within a new vision module. The vision module consists of two distinct concepts:

* Vision enhancement providers: providers of functionality that change the contents of the screen. A vision enhancement provider can be a magnifier, highlighter, or whatever else that changes the screen contents.

	This pr contains a highlighter based on a transparent overlay window on which lines are drawn around the several objects on screen.

	The [Screen Curtain add-on](https://addons.nvda-project.org/addons/screenCurtain.en.html), which is a prototype add-on written by me on behalf of @babbagecom and ported to an add-on by @josephsl, is supposed to be added to NVDA in subsequent pr.

	Vision enhancement providers will also be able to benefit from the functionality to set driver/provider specific options in the gui (#8214).

* The vision handler, which receives information about focus, foreground, navigator object, review and mouse changes from the core of NVDA. It keeps track of the location of objects on screen and communicates this information to one or more active vision enhancement providers. This information is communicated to vision enhancement providers using extension points. The handler keeps track of active enhancement providers and can activate and deactivate them. Therefore, it is very much similar to the braille handler in its approach.

### Follow up strategy
@feerrenrut has requested for this framework to be filed into subsequent pull requests. After this pr has been fully reviewed and merged, I intent to file at least two additional pr's:

1. A graphical interface for the framework
2. A built-in screen curtain

### Known issues
The current highlighter colors might be difficult for some people to see. In the future, I'd like this to be configurable.

### Changelog entries:
* Changes for developers
    + A vision framework has been added to NVDA. It allows you to create vision enhancement providers; modules that can change screen contents, optionally based on input from NVDA about object locations.
        * Add-ons can bundle their own providers in a visionEnhancementProviders folder.
        * See the vision and visionEnhancementProviders modules for the implementation of the framework and examples, respectively.

Commit:
Vision module without gui, with support for highlighting based on a wx window

Adriani90 commented:
Support from @nishimotz is very welcome I think. As part of the japanese NVDA Team, he is the creator of the add-on focus highlight
https://github.com/nvdajp/focusHighlight
I am sure that the japanese NVDA Team can bring much knowledge in this work as well.

nishimotz commented:
I am testing the AppVeyor build (nvda_snapshot_pr9064-16441,7aaee3c0.exe)
and confirmed the default highlighter works if nvda.ini is modified something like this

```
[vision]
        highlighter = NVDAHighlighter
        [[NVDAHighlighter]]
                highlightFocus = true
                highlightCaret = true
                highlightNavigatorObj = true
```

nishimotz commented on source/vision/__init__.py:
```diff
@@ -0,0 +1,701 @@
+		except NotImplementedError:
+			rects = None
+		if rects:
+			index = 0 if self.obj.isTextSelectionAnchoredAtStart else -1
```
I got errors as follows regarding this line:
```
ERROR - queueHandler.flushQueue (13:47:02.601):
Error in func _loadBufferDone from eventQueue
Traceback (most recent call last):
  File "queueHandler.pyc", line 53, in flushQueue
  File "virtualBuffers\__init__.pyc", line 464, in _loadBufferDone
  File "browseMode.pyc", line 1152, in event_treeInterceptor_gainFocus
  File "browseMode.pyc", line 1465, in event_gainFocus
  File "vision\__init__.pyc", line 601, in handleGainFocus
  File "vision\NVDAHighlighter.pyc", line 63, in updateContextRect
  File "vision\__init__.pyc", line 270, in updateContextRect
  File "vision\__init__.pyc", line 190, in getContextRect
  File "vision\__init__.pyc", line 205, in _getRectFromTextInfo
NameError: global name 'self' is not defined
```

Commit:
Fix bug reported by @nishimotz

Commit:
vision.VisionHandler.handleUpdate, treat updates to focus objects as a gainFocus, since the updated object might have a caret in it and we also want to account for this

Commit:
Log error when terminating a provider fails

Commit:
Make initializedProviders a set, since providers should be listed only once

Commit:
Make the virtual browse mode caret a bounding box

Commit:
Use visionEnhancementProviders package instead of class registration

Commit:
Merge remote-tracking branch 'origin/master' into visionbase

Commit:
make getProvider a public method on the vision module

Commit:
Fix highlighter according to new changes

nishimotz commented:
nvda_snapshot_pr9064-16574,1ca723ff.exe fails initialization as follows:

```
INFO - core.main (20:56:31.969):
NVDA version pr9064-16574,1ca723ff
INFO - core.main (20:56:31.969):
Using Windows version 10.0.17134 workstation

INFO - core.main (20:56:33.881):
NVDA initialized
ERROR - unhandled exception (20:56:33.881):
Traceback (most recent call last):
  File "wx\core.pyc", line 3240, in <lambda>
  File "vision.pyc", line 545, in setProvider
  File "vision.pyc", line 440, in getProvider
ImportError: No module named NVDAHighlighter
```

I cannot import this via Python console as well:

```
>>> import visionEnhancementProviders.NVDAHighlighter
Traceback (most recent call last):
  File "<console>", line 1, in <module>
ImportError: No module named NVDAHighlighter
```

Commit:
Make sure there is a visionEnhancementProviders module

leonardder commented:
Thanks @nishimotz! This should have been fixed in the most recent code.

Commit:
When looking up scripts, place visionEnhancementProviders higher up in the chain. We don't want random global plugins to override vision enhancement provider gesture people might rely on.
Also update some copyright headers

Commit:
Merge remote-tracking branch 'origin/master' into visionBase, fixing merge conflict

Commit:
Highlighter: rename ContextStyle to HighlightStyle and create a proper doc string

Commit:
Increase the width of the caret's rectangle

Commit:
Merge remote-tracking branch 'origin/master' into visionBase, fixing merge conflict and adding vision enhancement providers to scratchpad.

Commit:
Hopefully fixed issue resulting in the highlighter not willing to start due to an error caused by locationHelper conversion

Commit:
Use a AutoBufferedPaintDC for the highlighter

Commit:
Remove duplicated imports

Commit:
Merge remote-tracking branch 'origin/master' into visionBase

leonardder commented:
While the highlighter based on WX is doing its job pretty well in daily use, a major eyesore is that it's being hosted in NVDA's main thread. I've seen cases where the system freezes entirely when the log viewer is opened with the highlighter on.

I guess the only way to avoid this, is by using a separate thread for the highlight, like Focus Highlight also does. However, I guess that also means we have to abandon wx for this. Thoughts @nishimotz?

nishimotz commented:
I'm sorry to know that the highlighter working in the main thread is difficult.
Is there no option to use WX for drawing in a separate thread?

Another concern is that the current version of Focus Highlight uses GDI Plus because it seems easier to handle multiple monitors at physical coordinates.

leonardder commented:
> Is there no option to use WX for drawing in a separate thread?

I'm afraid we can't even create a window on a separate thread, but I'll do some further investigation.

> Another concern is that the current version of Focus Highlight uses GDI Plus because it seems easier to handle multiple monitors at physical coordinates.

Are you saying this is a concern for focus highlight, or for the wx based highlighter? Actually, apart from focus highlight not written very object orientedly, I believe it does it job quite well.

Commit:
Convert the highlighter to GDIPlus

Commit:
Do not cache properties by default on the vision handler, as it causes initialFocus to fail for the first loaded provider

Commit:
DO not force a refresh"

Commit:
Try to fix lag in highlighter

Commit:
Fix mouse flickering

Commit:
Merge remote-tracking branch 'origin/master' into visionBase

Commit:
Add visionEnhancementProviders to scratchpad only dirs

leonardder commented:
nishimotz: I updated the highlighter to use GDI Plus, based on your work but a bit more pythonic. I wonder what you think about the current approach. Also, tests would be greatly appreciated.

nishimotz commented:
Found an issue with nvda_snapshot_pr9064-17053,91896a0a.exe.

steps:

* create nvda snapshot as portable copy
* modify userConfig/nvda.ini for enabling vision framework
* start nvda
* start Chrome and visit a page, move focus to a link, for example
* press NVDA+space twice
* an error sound

```
ERROR - scriptHandler.executeScript (14:36:13.503):
error executing script: <bound method GlobalCommands.script_toggleVirtualBufferPassThrough of <globalCommands.GlobalCommands object at 0x04CB77D0>> with gesture u'NVDA+\u30b9\u30da\u30fc\u30b9'
Traceback (most recent call last):
  File "scriptHandler.pyc", line 196, in executeScript
  File "globalCommands.pyc", line 1261, in script_toggleVirtualBufferPassThrough
  File "treeInterceptorHandler.pyc", line 141, in _set_passThrough
  File "vision.pyc", line 630, in handleGainFocus
  File "visionEnhancementProviders\NVDAHighlighter.pyc", line 107, in updateContextRect
  File "vision.pyc", line 267, in updateContextRect
  File "vision.pyc", line 188, in getContextRect
AttributeError: 'ChromeVBuf' object has no attribute 'location'
```

So far, the highlight works nicely, as well as the reviewing character position.

Commit:
Fix passThrough failure

leonardder commented:
Thanks for reporting this. This should now be fixed.

How do you perceive the performance of this highlighter in comparison 
with early code in this pr and your own focus highlight add-on?

dingpengyu commented:
hi leonardder
Can you fix Failure: AppVeyor was unable to build non-mergeable pull request?

Commit:
Merge remote-tracking branch 'origin/master' into visionBase, fixing merge conflicts

leonardder commented:
Thanks for pointing this out!

nishimotz commented:
Confirmed that the Chrome passThrough issue is addressed.

I have found the difference between the results of vision framework and that of my add-on.

tested version: nvda_snapshot_pr9064-17078,432ec1ef.exe

steps:

* launch NVDA
* NVDA+N to open the menu
* press C to open the configuration profiles dialog
* press NVDA+B and wait until the speech is finished
* press NVDA+Shift+O (laptop) or NVDA+numpad 5 (desktop)
* it can be confirmed that the navigator object is the close button (right bottom)

The focus highlight add-on can show:
* focus is on the list item
* navigator object is the close button

Vision framework can show:
* only the focus, which is on the list item

leonardder commented:
I haven't looked at the code yet. However, this might be because focus 
highlight add-on continuously polls the navigator object, while the 
vision framework is event driven, i.e. if handleReviewMove isn't called 
explicitly, the vision framework doesn't know about the new navigator 
object.


Actually, I'm not sure whether this is a bug. The script help for NVDA+b 
clearly states that it speaks the window. I don't think the new 
navigator object is reflected in braille either, though I haven't yet 
tested this. Definitely something I should look into.

Commit:
Make custom window proc errors less obnoxious

Commit:
Make sure a possible magnifier gets hold of the location of the currently selected cell within Excel

Commit:
When getting a bounding rectangle from an offset in a virtual buffer (generic implementation), make sure a LookupError is raised when the object at a particular offset died

Commit:
Improve Python3 compatibility by no longer using the iter methods

Commit:
Fix typo in virtualBuffers package

Commit:
Added claryfing comments to doc strings

Commit:
Merge remote-tracking branch 'origin/master' into visionBase

dingpengyu commented:
hello leonardder
I noticed that this pr is not categorized in the AfterPython3Transition tag. Does this indicate that it might be merged into the master? Also, what problem is there still not solved in this pr?

Commit:
Convert vision enhancement providers to be based on driverHandler.Driver

Commit:
Rather than having one context for physical and virtual caret changes, split them in CONTEXT_CARET and CONTEXT_BROWSEMODE

Commit:
Use factory methods

Commit:
Remove extra tab

Commit:
Merge remote-tracking branch 'origin/master' into visionBase

Commit:
Prepare to make vision a package

Commit:
Merge remote-tracking branch 'origin/master' into visionBase

Commit:
Remove conflicting roles

Commit:
VisionEnhancementProvider.getContextObject, properly return an object when getting context object for caret and browse mode

Commit:
Remove enabled and running property from VisionEnhancementProvider

Commit:
Let highlighter and magnifier deal with disabled or unimplemented contexts themselves, i.e. don't leave this up to the handler

Commit:
Fix indentation in handleGainFocus

Commit:
Convert vision module to package with several submodules

Commit:
Wrap SetTimer and KillTimer into an object

Commit:
Merge remote-tracking branch 'origin/master' into visionBase

Commit:
Do not longer join the highlighter thread for now, as it could lock up NVDA

Commit:
providerBase.py should know about textInfos

Commit:
Fix case where information for the caret was silently discarded in favor of information about the selection

Commit:
Fix ColorTransformationInfo

Commit:
Fix getProvider

Commit:
Identify bare exceptions, code cleanup

Commit:
Make setProvider expect a sequence

Commit:
Hopefully fix bug with search button in GMail

Commit:
NVDAhighlighter: remove unneeded code, fixes

Commit:
CONTEXT_BROWSEMODe > CONTEXT_BROWSEMODE

Commit:
Add pkgutil to vision __init__

Commit:
Add check method to highlighter

Commit:
Remove exception for excel

Commit:
Excel7: use the focusRedirect system instead of event_gainFocus

Commit:
Excel7Window, let caret moves update the vision framework

Commit:
catch exceptions for initialFocus

nishimotz commented:
nvda_snapshot_pr9064-17306,32d9232f.exe raises errors as follows.

Should I modify nvda.ini?

nvda.log

```
INFO - __main__ (12:38:48.187):
Starting NVDA
INFO - core.main (12:38:49.013):
Config dir: C:\Users\nishimotz\AppData\Roaming\nvda
INFO - config.ConfigManager._loadConfig (12:38:49.013):
Loading config: C:\Users\nishimotz\AppData\Roaming\nvda\nvda.ini
INFO - core.main (12:38:49.030):
Developer Scratchpad mode enabled
INFO - core.main (12:38:49.086):
NVDA version pr9064-17306,32d9232f


ERROR - windowUtils._rawWindowProc (12:38:50.180):
Error in wndProc
Traceback (most recent call last):
  File "windowUtils.pyc", line 253, in _rawWindowProc
  File "visionEnhancementProviders\NVDAHighlighter.pyc", line 188, in windowProc
  File "visionEnhancementProviders\NVDAHighlighter.pyc", line 234, in _paint
AttributeError: 'HighlightWindow' object has no attribute 'location'


ERROR - eventHandler.executeEvent (12:38:50.513):
error executing event: gainFocus on <NVDAObjects.Dynamic_IAccessibleRichEdit50WindowNVDAObject object at 0x049E16B0> with extra args of {}
Traceback (most recent call last):
  File "eventHandler.pyc", line 155, in executeEvent
  File "eventHandler.pyc", line 92, in __init__
  File "eventHandler.pyc", line 100, in next
  File "appModules\nvda.pyc", line 86, in event_gainFocus
  File "eventHandler.pyc", line 100, in next
  File "NVDAObjects\behaviors.pyc", line 179, in event_gainFocus
  File "NVDAObjects\__init__.pyc", line 1036, in event_gainFocus
  File "vision\visionHandler.pyc", line 206, in handleGainFocus
NameError: global name 'mightHaveCaret' is not defined
```


vision group of nvda.ini

```
[vision]
	magnifier = None
	highlighter = NVDAHighlighter
	colorEnhancer = None
	[[NVDAHighlighter]]
		highlightFocus = True
		highlightCaret = True
		highlightNavigatorObj = True
```

Commit:
Restore location property on highlighter

This reverts commit 8be41d6e7bd368e931eea8b3e651312eb58ad48c.

feerrenrut commented:
This re-organisation looks a lot better to me. I think it would be worth adding a readme.md to the vision package. This can be helpful to explaing the layout a little, but also the concepts. I'm visualizing help and examples for add-on developers eventually as well. Perhaps this should be in the docscring for the package. Though for indepth information this could make the file mostly comments. Happy to hear opinions on where this should live.

leonardder commented:
There are more github projects using per directory readme.md files, yet 
it makes inspection of the package from the python console a whole lot 
more difficult, as the information is not available using that method.

nishimotz commented:
nvda_snapshot_pr9064-17312,c3faf25e.exe still raises errors as follows.

```
ERROR - eventHandler.executeEvent (11:22:22.559):
error executing event: gainFocus on <NVDAObjects.Dynamic_IAccessibleRichEdit50WindowNVDAObject object at 0x04BEE530> with extra args of {}
Traceback (most recent call last):
  File "eventHandler.pyc", line 173, in executeEvent
  File "eventHandler.pyc", line 110, in __init__
  File "eventHandler.pyc", line 118, in next
  File "appModules\nvda.pyc", line 86, in event_gainFocus
  File "eventHandler.pyc", line 118, in next
  File "NVDAObjects\behaviors.pyc", line 179, in event_gainFocus
  File "NVDAObjects\__init__.pyc", line 1036, in event_gainFocus
  File "vision\visionHandler.pyc", line 206, in handleGainFocus
NameError: global name 'mightHaveCaret' is not defined
```

Commit:
hasNavigableText property instead of mightHaveCaret

feerrenrut committed:
Rename getProvider to getProviderClass

nishimotz commented:
Tested nvda_snapshot_pr9064-17320,48bc1b0b.exe.
Very stable and reliable.

When Google Chrome opens a website, switch to browse mode, and press H to go to the first heading, the navigator object is shown as the first character of the heading element.

Is this intentional? 

Vision framework:

![image](https://user-images.githubusercontent.com/76300/58265283-4fef9080-7dba-11e9-8059-ffd77f4e1736.png)

Focus Highlight:

![image](https://user-images.githubusercontent.com/76300/58265863-78c45580-7dbb-11e9-89a9-5db72c69031c.png)

leonardder commented:
If everything is working as expected, you should see a box around the 
navigator object, and a box around the first character of the heading 
that highlights the currently selected character in browse mode. The 
latter box should change its position when you move with the left and 
right arrow keys from character to character.

k-kolev1985 commented:
Sorry for the maybe stupid question, but is there a way for me to test this? Is it implemented as a hidden feature in current "alpha" snapshots?

Commit:
Merge remote-tracking branch 'origin/master' into visionBase

Commit:
First work on rewrite of the handler, without roles

Commit:
More work on code restructure

Commit:
Change handleUpdate to have its own extension point

Commit:
Move highlighter logic to NVDAHighlighter

Commit:
Merge remote-tracking branch 'origin/master' into visionBase

Commit:
New style contexts

Commit:
Many updates to highlighter

Commit:
Remove magnifier base

leonardder commented:
@feerrenrut I think I have applied all the changes that we discussed by voice earlier. Could you have another look?

Commit:
Small improvements

Commit:
Add a wx.App instance to tests

feerrenrut commented on source/vision/visionHandler.py:
```diff
@@ -0,0 +1,194 @@
+		# However, execute this on the main thread, which makes sure that
+		# the gui is fully initialized before providers are initialized that might rely on it.
+		wx.CallAfter(self.handleConfigProfileSwitch)
+		config.post_configProfileSwitch.register(self.handleConfigProfileSwitch)
```
Could `post_configProfileSwitch` be triggered before the `wx.CallAfter`, short-circuiting this logic. Perhaps instead:

```
		wx.CallAfter(self.postGuiInit)

	def postGuiInit(self):
		self.handleConfigProfileSwitch
		config.post_configProfileSwitch.register(self.handleConfigProfileSwitch)
```

feerrenrut commented on source/vision/visionHandler.py:
```diff
@@ -0,0 +1,194 @@
+			# A provider can raise whatever exception,
+			# therefore it is unknown what to expect.
+			log.error("Error while terminating vision provider %s" % providerName, exc_info=True)
+			return False
```
Beware, this leaves the `providers` list missing `providerInstance`, but returns indicating that it can't be removed. The other `return False` has the same issue. This may be intentional, in which case I think it deserves a comment about why it's what we want.

feerrenrut commented on source/vision/visionHandler.py:
```diff
@@ -0,0 +1,194 @@
+		wx.CallAfter(self.handleConfigProfileSwitch)
+		config.post_configProfileSwitch.register(self.handleConfigProfileSwitch)
+
+	def terminateProvider(self, providerName):
```
What does the return mean? I assume successfully terminated, what does an unsuccessful termination mean?

feerrenrut commented on source/vision/visionHandler.py:
```diff
@@ -0,0 +1,194 @@
+			log.error("Error while terminating vision provider %s" % providerName, exc_info=True)
+			return False
+		# Copy the configured providers before mutating the list
+		configuredProviders = config.conf['vision']['providers'][:]
```
This is a handy trick.

feerrenrut commented on source/vision/visionHandler.py:
```diff
@@ -0,0 +1,194 @@
+			config.conf['vision']['providers'] = configuredProviders
+		except ValueError:
+			pass
+		# Recreate our extension points instance
```
I think this deserves an explanation of "why": "During termination of a provider, the instance is not destroyed (for efficiency in slow to start providers), we cant rely on providers to de-register themselves, so instead we recreate the extensionPoints and ask active providers to re-register."

feerrenrut commented on source/vision/providerBase.py:
```diff
@@ -0,0 +1,47 @@
+	def registerEventExtensionPoints(self, extensionPoints):
+		"""
+		Called at provider initialization time, this method should register the provider
+		to the several event extension points that it is interested in.
```
I would expand this comment to mention that it should only register with the extension points, it may be called at several times between initialisation and termination.

feerrenrut commented on source/vision/visionHandlerExtensionPoints.py:
```diff
@@ -0,0 +1,25 @@
+	post_mouseMove = None
+
+	def __init__(self):
+		self.post_focusChange = Action()
```
Some docs here, especially to document the parameters passed in.

feerrenrut commented on source/visionEnhancementProviders/NVDAHighlighter.py:
```diff
@@ -0,0 +1,300 @@
+
+import vision
+from vision.constants import Role, Context
+from vision.util import *
```
Please specify the functions used, or import as something.

feerrenrut commented on source/visionEnhancementProviders/NVDAHighlighter.py:
```diff
@@ -0,0 +1,300 @@
+"""Default highlighter based on GDI Plus."""
+
+import vision
+from vision.constants import Role, Context
```
these aren't used

feerrenrut commented on source/visionEnhancementProviders/NVDAHighlighter.py:
```diff
@@ -0,0 +1,300 @@
+from logHandler import log
+from mouseHandler import getTotalWidthAndHeightAndMinimumPosition
+from locationHelper import RectLTWH
+import config
```
not used

feerrenrut commented on source/visionEnhancementProviders/NVDAHighlighter.py:
```diff
@@ -0,0 +1,300 @@
+	@ivar color: THe color to use for the style
+	@type color: L{RGB}
+	@ivar width: The width of the lines to be drawn, in pixels.
+		A higher width reduces the inner dimentions of the rectangle.
```
```suggestion
		A higher width reduces the inner dimensions of the rectangle.
```

feerrenrut commented on source/visionEnhancementProviders/NVDAHighlighter.py:
```diff
@@ -0,0 +1,300 @@
+	@type style: int
+	@ivar margin: The number of pixels between the highlight's rectangle
+		and the rectangle of the object to be highlighted.
+		A higher margin increases both the inner and outer dimentions of the highlight's rectangle.
```
```suggestion
		A higher margin increases both the inner and outer dimensions of the highlight's rectangle.
```

feerrenrut commented on source/visionEnhancementProviders/NVDAHighlighter.py:
```diff
@@ -0,0 +1,300 @@
+	@type color: L{RGB}
+	@ivar width: The width of the lines to be drawn, in pixels.
+		A higher width reduces the inner dimentions of the rectangle.
+		Therefore, if you need to increase the outer dimentions of the rectangle, you need to increase the margin as well.
```
```suggestion
		Therefore, if you need to increase the outer dimensions of the rectangle, you need to increase the margin as well.
```

feerrenrut commented on source/visionEnhancementProviders/NVDAHighlighter.py:
```diff
@@ -0,0 +1,300 @@
+	description = _("NVDA Highlighter")
+	supportedContexts = (Context.FOCUS, Context.NAVIGATOR, Context.BROWSEMODE)
+	_ContextStyles = {
+		Context.FOCUS: HighlightStyle(RGB(0x03, 0x36, 0xff), 5, winGDI.DashStyleDash, 5),
```
Please pull these colours and styles out into constants.

EG:
```
BLUE = RGB(0x03, 0x36, 0xff)
PINK = RGB(0xff, 0x02, 0x66)
YELLOW = RGB(0xff, 0xde, 0x03)

DASH_BLUE =  HighlightStyle(BLUE, 5, winGDI.DashStyleDash, 5)
SOLID_PINK = HighlightStyle(PINK, 5, winGDI.DashStyleSolid, 5)
SOLID_BLUE = HighlightStyle(BLUE, 5, winGDI.DashStyleSolid, 5)
SOLID_YELLOW = HighlightStyle(YELLOW, 2, winGDI.DashStyleSolid, 2)
```

feerrenrut commented on source/visionEnhancementProviders/NVDAHighlighter.py:
```diff
@@ -0,0 +1,300 @@
+		This value may also be negative.
+	@type margin: int
+	"""
+	__slots__ = ()
```
Whats this for?

feerrenrut commented on source/vision/visionHandler.py:
```diff
@@ -0,0 +1,194 @@
+			"visionEnhancementProviders.%s" % moduleName,
+			globals(),
+			locals(),
+			("visionEnhancementProviders",)
```
I think `__import__` expects a list for this arg, but gets a tuple.

feerrenrut reviewed (changes requested):
This is looking good to me, lots of small comments, but I expect this will be good to go after this.

Commit:
Give appModules and globalPlugins precedence over vision enhancement providers when finding a script for a particular gesture. Furthermore, fixed a bug that excluded braille display drivers from the scripthandler.findScript function

Commit:
Many more doc strings and comments

Commit:
More review actions

leonardder commented on source/scriptHandler.py:
```diff
@@ -96,6 +98,19 @@ def findScript(gesture):
+		if func:
+			return func
+
+	# Vision enhancement provider level
```
I can't find your initial comment about this, but I revisited this. Now, global plugins and appModules take precedence over vision enhancement provider scripts. I think this makes more sense than it was before.

leonardder commented on source/vision/__init__.py:
```diff
@@ -0,0 +1,77 @@
+			)
+			continue
+		try:
+			if not onlyStartable or provider.check():
```
You had a comment about using the word check as a name for the function. I'm very reluctant to change this as in fact, check is a method on driverHandler.Driver. In other words, check is used this way for BrailleDisplayDrivers and SynthDrivers as well. I think I prefer consistency over clarity here.

feerrenrut replied:
Consistently unclear? I missed that it's on the `Driver` base class.

This is an opportunity to adapt the interface, and provide further clarity. Consider the following psudo-code approach:

``` Python
class VisionEnhancementProvider(driverHandler.driver):

	@abstractclassmethod
	def canStart():
		raise NotImplementedError

	def check():
		return self.canStart()

```
You might even want to re-map `check` to `canStart` in `__new__` or `__init__` if you wanted to save the extra function call overhead, but I doubt the performance cost justifies the cost to readability.

leonardder replied:
Unfortunately, there is no such thing as an abstract class method.

HOw about implementing canStart for braille and speech as well?

Then:

* We rename check to canStart
* Calling check  raises a deprecation warning
* The base method of canStart calls check on classes where check is implemented/available.

leonardder replied:
May be isAvailable is a better name then canStart? see the doc string of check on driverHandler.Driver

feerrenrut replied:
A [stackoverflow answer](https://stackoverflow.com/a/11218474/346188) presents a solution for an abstract class method. This won't stop you from calling 
`VisionEnhancementProvider.canStart()`, you need `raise NotImplementedError` for that, but it will stop you from creating an instance of the class that does not override the class method. It also communicates the intent quite well.

> HOw about implementing canStart for braille and speech as well?

I think I would prefer not to extend the scope of the PR that far, it would logically follow from there that we get rid of `check` entirely. 

> May be isAvailable is a better name then canStart? see the doc string of check on driverHandler.Driver

Perhaps, but only if this really fits what we want the providers to tell us, this method should be adapting the generic `Driver.check` function, to make more sense in the context of a `VisionEnhancementProvider`. I think we want them to tell us more than "are they available", but also if they can actually be started / run. Say two providers require the same resources, they can't both run at once, this is their chance to say they are not able to. Am I understanding the design correctly?

leonardder commented on source/vision/util.py:
```diff
@@ -0,0 +1,88 @@
+import api
+import locationHelper
+
+def getReviewRect():
```
This was getReviewRectangle before. I decided to rename these in order to improve consistency.

leonardder commented:
@feerrenrut: Thanks for your extensive review! I think I've addressed all the comment, particularly by adding many comments and additional doc strings.

josephsl commented:
Hi, one small favor: can you test if this is Python 3 compliant for future-proofing purposes? Thanks.

 

From: Leonard de Ruijter <notifications@github.com> 
Sent: Tuesday, July 2, 2019 7:09 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [nvaccess/nvda] Abstract vision framework with included support for focus, navigator object and browse mode caret highlighting (#9064)

 

@feerrenrut <https://github.com/feerrenrut> : Thanks for your extensive review! I think I've addressed all the comment, particularly by adding many comments and additional doc strings.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/9064?email_source=notifications&email_token=AB4AXECY7PUS37P4SQ5EXVLP5NOONA5CNFSM4GKG6Y3KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZBMTHA#issuecomment-507693468> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AB4AXEGMTTPP2KE2SR7ORZTP5NOONANCNFSM4GKG6Y3A> .

Commit:
Test suite: make sure we import sourceEnv so the tests are aware of all our dependencies

leonardder commented:
I've tried to ensure compatibility, but as I proposed elsewhere, I think 
we should do the merge of this into threshold by hand, preferably after 
threshold_py3_staging went into threshold.

Commit:
Merge remote-tracking branch 'origin/master' into visionBase

Commit:
Add context for foreground

Commit:
Add vision to the python console namespace

feerrenrut commented on source/vision/visionHandlerExtensionPoints.py:
```diff
@@ -4,21 +4,105 @@
+	#: For example, a magnifier can track the magnified area of the screen to the new caret position.
+	#: Handlers are strongly encouraged to cache the new caret position,
+	#: and handle the pending caret update at the end of every core cycle using L{post_coreCycle},
+	#: unless tehy delegate caret change handling to a separate thread.
```
minor typo: tehy -> they

feerrenrut replied:
This typo looks like it has been copied to a few of these comments, a search and replace should fix this.

feerrenrut commented on source/vision/constants.py:
```diff
@@ -28,7 +28,12 @@ class Role:
 class Context:
+	"""context for events received by providers.
+	Typically this informs of the cause of the event.
+	For example, L{focus} is used when an event is trigggered by the focus object or a focus change.
```
minor typo: trigggered -> triggered

Commit:
Fix bad copy paste typo

Commit:
minor typo: trigggered -> triggered

feerrenrut commented:
> we should do the merge of this into threshold by hand, preferably after threshold_py3_staging went into threshold.

Yes, this makes sense. I think it's best to merge the py 3 changes into this when you are ready, commit that even with errors, then commit fixes for any errors on top. You can wait for threshold to have py 3 or merge threshold_py3_staging, I won't squash merge to threshold, I think the history will be useful

leonardder commented:
I will wait for threshold to have threshold_py3_staging. It might even 
be best to wait until master is py3, as otherwise we yet might get merge 
conflicts.

Commit:
Convert constants to enums

Commit:
Extension points: use data class

Commit:
Add many type hints to vision framework

Commit:
Fix small mistakes in vision.util

Commit:
Additional logging in vision.util

Commit:
Implement canStart abstract class method

Commit:
Remove sourceEnv from tests

Commit:
supportedSettings to class level

Commit:
Remove accidentally duplicated line

Commit:
If a provider fails to terminate, we yet want to update the config.

Commit:
Import wx after importing sourceEnv

Commit:
Set the transparent brush on the window class to make it clear the window automatically

Commit:
Change code order in highlighter module to be able to set the window class as a class property on the highlighter provider

Commit:
Use a constant for the transparency color

Commit:
Fix some small bugs, particularly related to browse mode caret following

Commit:
Make sure that review move is also executed when changing focus and becomeNavigatorObject is executed

Commit:
Catch TypeError when getting context rectangle in highlighter

Commit:
Linting actions

Commit:
Merge remote-tracking branch 'origin/master' into visionBase

Commit:
fixes to winGdi module to ensure Python 3 compatibility

leonardder commented:
@feerrenrut: I've tried up this branch a bit after the Python 3 merge. At least all relevant new commits since your last view of this pr are now one after the other.
Particularly python 3 changes have been applied.

Commit:
Merge remote-tracking branch 'origin/master' into visionBase

Commit:
Linter actions

Commit:
Merge remote-tracking branch 'origin/master' into visionBase

Commit:
More fixes for linting, now passes

feerrenrut reviewed (approved):
Ok looks good

feerrenrut commented:
Except there is a failing unit test, "No module named 'wx'"

leonardder commented:
Ah, I'm sorry for that. This all has to do with the linter complaining about module imports have to be at the top of the file.

Commit:
Another attempt to work around unit test framework linting errors

leonardder commented:
@feerrenrut I'm afraid I had to use some noqa in the unit module to get this work. This is because we have to change the path before importing modules, and this is something Flake8 doesn't agree with.

The sourceEnv noqa is intentional, as this module is only imported to manipulate the path.

feerrenrut commented on tests/unit/__init__.py:
```diff
@@ -33,7 +31,9 @@
 # Let us import modules from the NVDA source.
 sys.path.insert(1, SOURCE_DIR)
-import sourceEnv
+import sourceEnv  # noqa: F401
```
I'm fine with using `noqa` to suppress this case, however please include a comment about why and a hint about what the lint error is. It's not much fun for a future developer to have to look these up and understand why it is like this.

leonardder replied:
What would be proper syntax for this? I assume something like this is not going to work

```suggestion
import sourceEnv  # noqa: F401  module imported but unused
```

feerrenrut replied:
Yep something like that works, though it gets messy if you have multiple errors to ignore. 

I would do:
```
# F401: module imported but unused. We need sourceEnv for unit tests because of ?
import sourceEnv  # noqa: F401
```

leonardder replied:
I think I've done something along those lines.

Commit:
Add explanatory comments about noqa. Also ignore E402 instead of F402

leonardder commented:
Mm, I was ignoring linter warning F402, which had to be E402. Interesting enough, Ignoring F402 also ignored E402 unexpectedly, that's why I didn't see this before.

Commit:
Improve exception handling for vision.handler.initializeProvider

Commit:
Linter fix

feerrenrut commented on tests/unit/__init__.py:
```diff
@@ -70,7 +75,13 @@ class AppArgs:
 import appModuleHandler
 appModuleHandler.initialize()
-# Anything which notifies of cursor updates requires braille to be initialized.
+# Vision needs a wx app to be available.
```
From a unit testing perspective it would be better if we can avoid this. Which part of the vision framework relies on a wx app being available? Can we mock this part instead?

feerrenrut reviewed (approved):
Generally this looks good. But I am a little worried about the dependencies of vision, if possible I would prefer it rely on dependency injection if possible. Can we start by identifying what fails if wx is missing?

leonardder commented:
wx is used to make sure that the gui is initialized (CallAfter) before initializing providers. This, because providers might rely on the gui. The highlit"ghter, for example, relies on gui.mainFrame.handle to be the parent of its window.
By default, there is no provider active. So it might make sense to check for configured providers before using wx.CallAfter. If we ever decide to write a pseudo provider, this requires another revisit.
We can also mock wx.CallAfter, e.g. 'lambda func, *args, **kwargs: func(*args, **kwargs)'. Other code might benefit from this too.

Commit:
No longer rely on wx for the unit testing framework

Commit:
windowUtils improvements:

* CustomWindow.className should be abstract, which is now possible in Python 3.
* Ensure that when calling del on a window, handle the case where handle is not yet set on the window
* Add additional logging to windowUtils Window destruction.
* Use the typing module

Commit:
Fixes to highlighter:

* In Python 2, we couldn't join the highlighter thread, as it stopped the quit message from arriving. This now seems to work again.
* Reinitialization is no longer broken. We should create the transparent brush at window class creation time and not at class construction time.
